pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/LinguisticNamingRule.java

```diff
@@ -0,0 +1,148 @@
+            .desc("Check local variable names and types for inconsistent naming").uiOrder(7.0f).build();
+    private static final StringMultiProperty BOOLEAN_FIELD_PREFIXES_PROPERTY = StringMultiProperty
+            .named("booleanFieldPrefixes").defaultValues("is", "has", "can", "have", "will", "should")
+            .desc("the prefixes of fields that return boolean").uiOrder(8.0f).build();
+
+    public LinguisticNamingRule() {
+        definePropertyDescriptor(CHECK_BOOLEAN_METHODS);
+        definePropertyDescriptor(CHECK_GETTERS);
+        definePropertyDescriptor(CHECK_SETTERS);
+        definePropertyDescriptor(CHECK_TRANSFORM_METHODS);
+        definePropertyDescriptor(BOOLEAN_METHOD_PREFIXES_PROPERTY);
+        definePropertyDescriptor(CHECK_FIELDS);
+        definePropertyDescriptor(CHECK_VARIABLES);
+        definePropertyDescriptor(BOOLEAN_FIELD_PREFIXES_PROPERTY);
+        addRuleChainVisit(ASTMethodDeclaration.class);
+        addRuleChainVisit(ASTFieldDeclaration.class);
+        addRuleChainVisit(ASTLocalVariableDeclaration.class);
+    }
+
+    @Override
+    public Object visit(ASTMethodDeclaration node, Object data) {
+        String nameOfMethod = node.getMethodName();
+
+        if (getProperty(CHECK_BOOLEAN_METHODS)) {
+            checkBooleanMethods(node, data, nameOfMethod);
+        }
+
+        if (getProperty(CHECK_SETTERS)) {
+            checkSetters(node, data, nameOfMethod);
+        }
+
+        if (getProperty(CHECK_GETTERS)) {
+            checkGetters(node, data, nameOfMethod);
+        }
+
+        if (getProperty(CHECK_TRANSFORM_METHODS)) {
+            checkTransformMethods(node, data, nameOfMethod);
+        }
+
+        return data;
+    }
+
+    private void checkTransformMethods(ASTMethodDeclaration node, Object data, String nameOfMethod) {
+        ASTResultType resultType = node.getResultType();
+        if (nameOfMethod.contains("To") && resultType.isVoid()) {
+            // To in the middle somewhere
+            // a transform method shouldn't return void linguistically
+            addViolation(data, node);
+        } else if (hasPrefix(nameOfMethod, "to") && resultType.isVoid()) {
+            // a transform method shouldn't return void linguistically
+            addViolation(data, node);
+        }
+    }
+
+    private void checkGetters(ASTMethodDeclaration node, Object data, String nameOfMethod) {
+        ASTResultType resultType = node.getResultType();
+        if (hasPrefix(nameOfMethod, "get") && resultType.isVoid()) {
+            // get method shouldn't return void linguistically
+            addViolation(data, node);
+        }
+    }
+
+    private void checkSetters(ASTMethodDeclaration node, Object data, String nameOfMethod) {
+        ASTResultType resultType = node.getResultType();
+        if (hasPrefix(nameOfMethod, "set") && !resultType.isVoid()) {
+            // set method shouldn't return any type except void linguistically
+            addViolation(data, node);
```

Comment:
I think adding an explanation message for each case would be nicer than the current blanket message. It doesn't appear too hard to do, the comments could be promoted to a message argument

Reply:
Yes, that makes sense.